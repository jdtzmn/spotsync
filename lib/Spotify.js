#!/usr/bin/env node
var https = require('https');

var Spotify = function(cb) {

  //general request functions:

  var port = 4370;
  var csrf = '';
  var token = '';

  var request = function(path, cb) {
    var id = function() {
      return Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 10) + '.';
    };

    var options = {
      hostname: id() + 'spotilocal.com',
      port: port,
      path: path,
      method: 'GET',
      rejectUnauthorized: false,
      headers: {
        'Origin': 'https://embed.spotify.com'
      }
    };

    var req = https.request(options, function(res) {
      var response = '';
      res.on('data', function(d) {
        response += d;
      });

      res.on('end', function() {
        var data = JSON.parse(response);
        cb(data);
      });
    });
    req.end();

    req.on('error', function(err) {
      cb(err);
    });
  };

  //custom functions:

  this.ready = function ready(open, cb) {

    if (typeof open === 'function') cb = open;

    if (port >= 4379) return new Error('spotify web helper not running.');

    request('/service/version.json?service=remote', function(response) {
      if (response.code === 'ECONNREFUSED' && port < 4380) {
        port += 1;
        ready(cb);
      } else {

        request('/simplecsrf/token.json', function(response) {
          csrf = response.token;

          var req = https.request({
            hostname: 'open.spotify.com',
            method: 'GET',
            path: '/token'
          }, function(res) {
            var response = '';
            res.on('data', function(d) {
              response += d;
            });

            res.on('end', function() {
              var data = JSON.parse(response);
              token = data.t;
              if (open === true) {
                request('/remote/open.json', function(response) {
                  cb();
                });
              } else {
                cb();
              }
            });
          });
          req.end();

          req.on('error', function(e) {
            cb(e);
          });

        });
      }
    });
  };

  this.open = function open(cb) {
    request('/remote/open.json', function() {
      cb();
    });
  };

  this.status = function status(cb) {
    request('/remote/status.json?csrf=' + csrf + '&oauth=' + token, function(response) {
      cb(response);
    });
  };

  this.play = function play(uri, cb) {
    if (!uri || typeof uri !== 'string') {
      cb(new Error('uri is needed in order to play song.'));
    } else {
      request('/remote/play.json?csrf=' + csrf + '&oauth=' + token + '&uri=' + uri, function(response) {
        cb(response);
      });
    }
  };

  this.toggle = function toggle(cb) {
    this.status(function(data) {
      request('/remote/pause.json?csrf=' + csrf + '&oauth=' + token + '&pause=' + data.playing, function(response) {
        cb(response);
      });
    });
  };

  this.pause = function pause(cb) {
    request('/remote/pause.json?csrf=' + csrf + '&oauth=' + token + '&pause=true', function(response) {
      cb(response);
    });
  };
};

module.exports = Spotify;
