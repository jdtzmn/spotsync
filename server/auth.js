const { URL } = require('url')
const request = require('request')
const express = require('express')
const auth = express.Router()

module.exports = (clientId, clientSecret) => {
  // --------------------------------
  // ------------- LOGIN ------------
  // --------------------------------

  auth.get('/login', (req, res) => {
    const authURL = new URL('https://accounts.spotify.com/authorize')
    authURL.searchParams.append('client_id', clientId)
    authURL.searchParams.append('response_type', 'code')

    // generate scopes
    const scopes = ['streaming', 'user-read-birthdate', 'user-read-email', 'user-read-private', 'user-read-playback-state', 'user-modify-playback-state']
    authURL.searchParams.append('scope', scopes.join(' '))

    // generate redirect_uri
    let redirectURL = new URL(req.protocol + '://' + req.get('host'))
    redirectURL.pathname = '/auth/callback'
    authURL.searchParams.append('redirect_uri', redirectURL.toString())

    // generate and set state token
    let state = (Math.random().toString(36) + '00000000000000000').slice(2, 12)
    res.cookie('spotify_state', state, {
      expires: new Date(Date.now() + 1e4),
      signed: true,
      httpOnly: true
    })
    authURL.searchParams.append('state', state)

    // set origin cookie
    let origin = req.query.origin
    res.cookie('origin', origin, { httpOnly: true })

    // redirect the request
    res.redirect(authURL.toString())
  })

  // --------------------------------
  // ------- ACCEPT CALLBACK --------
  // --------------------------------
  auth.get('/callback', (req, res) => {
    const responseURL = new URL(req.protocol + '://' + req.get('host'))
    const origin = req.cookies ? req.cookies['origin'] : null
    res.clearCookie('origin')
    responseURL.pathname = decodeURIComponent(origin) || '/'

    const code = req.query.code
    const state = req.query.state
    const cookieState = req.signedCookies ? req.signedCookies['spotify_state'] : null

    if (state && state === cookieState) {
      const redirectURL = new URL(req.protocol + '://' + req.get('host'))
      redirectURL.pathname = '/auth/callback'
      const redirectURI = redirectURL.toString()

      const config = {
        url: 'https://accounts.spotify.com/api/token',
        form: {
          grant_type: 'authorization_code',
          code: code,
          redirect_uri: redirectURI
        },
        headers: {
          Authorization: 'Basic ' + (new Buffer(clientId + ':' + clientSecret).toString('base64'))
        },
        json: true
      }

      request.post(config, (err, response, body) => {
        if (err || response.statusCode !== 200) {
          responseURL.searchParams.append('error', 'invalid_token')
          res.redirect(responseURL.toString())
        } else {
          // set refresh token to expire in 2 weeks
          const refreshToken = body.refresh_token
          res.cookie('refresh_token', refreshToken, {
            expires: new Date(Date.now() + 12096e5),
            signed: true,
            httpOnly: true
          })

          const accessToken = body.access_token

          // set access token to expire in an hour
          res.cookie('access_token', accessToken, {
            expires: new Date(Date.now() + 36e5),
            signed: true,
            httpOnly: true
          })
          res.redirect(responseURL.toString())
        }
      })
    } else {
      responseURL.searchParams.append('error', 'state_mismatch')
      res.redirect(responseURL.toString())
    }
  })

  // --------------------------------
  // ---------- SEND TOKEN ----------
  // --------------------------------

  auth.post('/token', (req, res) => {
    if (req.signedCookies && req.signedCookies['access_token']) {
      res.send(req.signedCookies['access_token'])
    } else if (req.signedCookies && req.signedCookies['refresh_token']) {
      const config = {
        url: 'https://accounts.spotify.com/api/token',
        form: {
          grant_type: 'refresh_token',
          refresh_token: req.signedCookies['refresh_token']
        },
        headers: {
          Authorization: 'Basic ' + (new Buffer(clientId + ':' + clientSecret).toString('base64'))
        },
        json: true
      }

      request.post(config, (err, response, body) => {
        if (err || response.statusCode !== 200) {
          res.sendStatus(404)
        } else {
          const accessToken = body.access_token

          // set access token to expire in an hour
          res.cookie('access_token', accessToken, {
            expires: new Date(Date.now() + 36e5),
            signed: true,
            httpOnly: true
          })
          res.send(accessToken)
        }
      })
    } else {
      res.sendStatus(404)
    }
  })

  // --------------------------------
  // ------------- LOGOUT -----------
  // --------------------------------

  auth.get('/logout', (req, res) => res.clearCookie('access_token').clearCookie('refresh_token').redirect('/'))

  // -------- RETURN ROUTER ---------
  return auth
}
