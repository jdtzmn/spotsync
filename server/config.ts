// Register environment variables
import { config } from 'dotenv'
config()

export const cookieSecret = process.env.COOKIE_SECRET
export const scopes = 'user-modify-playback-state user-read-currently-playing user-read-playback-state'
export const clientId = process.env.SPOTIFY_CLIENT_ID
export const clientSecret = process.env.SPOTIFY_CLIENT_SECRET
export const redirectUri = process.env.SPOTIFY_REDIRECT_URI || 'http://localhost:3000/auth/redirect'
