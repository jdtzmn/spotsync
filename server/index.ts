import express from 'express'
import { Server } from 'http'
import next from 'next'
import bodyParser from 'body-parser'

import authRoutes from './auth'
import handleSockets from './ws'

const port = parseInt(process.env.PORT, 10) || 3000
const dev = process.env.NODE_ENV !== 'production'
const nextApp = next({ dev })
const handle = nextApp.getRequestHandler()

nextApp.prepare()
  .then(() => {
    const app = express()
    const server = new Server(app)

    /* ==================== */
    /* ====== SOCKETS ===== */
    /* ==================== */

    const io = require('socket.io')(server)
    handleSockets(io)

    /* ==================== */
    /* ====== ROUTES ====== */
    /* ==================== */

    app.use(bodyParser.json())

    app.use('/auth', authRoutes)

    app.get('*', (req, res) => {
      return handle(req, res)
    })

    server.listen(port, (err) => {
      if (err) throw err
      // tslint:disable no-console
      console.log(`> Ready on http://localhost:${port}`)
      // tslint:enable no-console
    })
  })
