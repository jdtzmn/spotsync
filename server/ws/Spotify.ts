import axios, { AxiosInstance, AxiosError, AxiosPromise } from 'axios'
import { stringify as urlEncode } from 'querystring'

export default class Spotify {
  private readonly clientId: string
  private readonly clientSecret: string
  private instance: AxiosInstance = null

  get isSetup () {
    return this.instance !== null
  }

  constructor(clientId: string, clientSecret: string) {
    this.clientId = clientId
    this.clientSecret = clientSecret
  }

  public setup () {
    return this.requestAccessToken()
      .then((accessToken) => {
        this.instance = axios.create({
          baseURL: 'https://api.spotify.com/v1/',
          headers: {
            Authorization: `Bearer ${accessToken}`
          }
        })
      })
  }

  public getTrack (uri: string) {
    const request = () => this.instance.get(`/tracks/${uri}`)
    return this.wrapRequest(request)
  }

  private requestAccessToken () {
    const body = { grant_type: 'client_credentials' }
    const authorization = `Basic ${Buffer.from(this.clientId + ':' + this.clientSecret).toString('base64')}`
    return axios.post('https://accounts.spotify.com/api/token', urlEncode(body), {
      headers: {
        Authorization: authorization
      }
    }).then((response) => response.data.access_token)
  }

  private async setupIfNotAlready () {
    if (!this.isSetup) await this.setup()
  }

  private async wrapRequest (request: () => AxiosPromise) {
    await this.setupIfNotAlready()

    try {
      const { data } = await request()
      return data
    } catch (error) {
      const { response: { status } } = error as AxiosError

      if (status === 401) {
        await this.setup()
        return this.wrapRequest(request)
      } else {
        throw error
      }
    }
  }
}
