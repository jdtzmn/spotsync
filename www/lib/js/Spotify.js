var Spotify = function(csrf) {

  var jquery = typeof jQuery !== 'undefined' ? true : false;

  //general request functions:

  var port = 4371;
  var token = '';

  var request = function(path, cb) {
    var id = function() {
      return Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 10) + '.';
    };

    if (jquery) {
      var options = {
        url: 'https://' + id() + 'spotilocal.com:' + port + path,
        error: function(xhr, status) {
          cb(status);
        },
        success: function(response) {
          var data = response;
          cb(data);
        }
      };

      $.get(options);
    } else {
      var req = new XMLHttpRequest();
      req.open('GET', 'https://' + id() + 'spotilocal.com:' + port + path, true);

      req.onload = function() {
        if (req.status >= 200 && req.status < 400) {
          var data = JSON.parse(req.responseText);
          cb(data);
        } else {
          cb('error');
        }
      };

      req.onerror = function() {
        cb('error');
      };

      req.send();
    }
  };

  //custom functions:

  this.ready = function ready(cb) {

    request('/service/version.json?service=remote', function(response) {
      if (response === 'error' && port < 4380) {
        port += 1;
        ready(cb);
      } else if (response === 'error' && port >= 4380) {
        cb("Spotify application is not running or doesn't support the internal web server.");
      } else {

        if (jquery) {
          $.ajax({
            url: 'https://jsonp.afeld.me/?url=https://open.spotify.com/token',
            method: 'GET',
            success: function(response) {
              token = response.t;
              request('/remote/status.json?csrf=' + csrf + '&oauth=' + token, function(response) {
                if (typeof response.error !== 'undefined' && response.error.type === "4107") {
                  cb('invalid CSRF token');
                } else {
                  cb();
                }
              });
            }
          });
        } else {
          var req = new XMLHttpRequest();
          req.open('GET', 'https://jsonp.afeld.me/?url=https://open.spotify.com/token', true);

          req.onload = function() {
            if (req.status >= 200 && req.status < 400) {
              var data = JSON.parse(req.responseText);
              token = data.t;
              request('/remote/status.json?csrf=' + csrf + '&oauth=' + token, function(response) {
                if (typeof response.error !== 'undefined' && response.error.type === "4107") {
                  cb('invalid CSRF token');
                } else {
                  cb();
                }
              });
            } else {
              cb('error');
            }
          };

          req.onerror = function() {
            cb('error');
          };

          req.send();
        }

      }
    });
  };

  this.status = function status(cb) {
    request('/remote/status.json?csrf=' + csrf + '&oauth=' + token, function(response) {
      cb(response);
    });
  };

  this.play = function play(uri, cb) {
    if (!uri || typeof uri !== 'string') {
      cb(new Error('uri is needed in order to play song.'));
    } else {
      request('/remote/play.json?csrf=' + csrf + '&oauth=' + token + '&uri=' + uri, function(response) {
        cb(response);
      });
    }
  };

  this.toggle = function toggle(cb) {
    this.status(function(data) {
      request('/remote/pause.json?csrf=' + csrf + '&oauth=' + token + '&pause=' + data.playing, function(response) {
        cb(response);
      });
    });
  };

  this.pause = function pause(cb) {
    request('/remote/pause.json?csrf=' + csrf + '&oauth=' + token + '&pause=true', function(response) {
      cb(response);
    });
  };
};
