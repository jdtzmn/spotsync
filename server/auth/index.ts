import { randomBytes } from 'crypto'
import { stringify } from 'querystring'
import { URL } from 'url'
import express from 'express'
import cookieParser from 'cookie-parser'
import { celebrate, Joi, errors } from 'celebrate'
import axios from 'axios'

/* ==================== */
/* ======= SETUP ====== */
/* ==================== */

// Config
import { cookieSecret, scopes, clientId, clientSecret, redirectUri } from '../config'
const authorizationUrl = 'https://accounts.spotify.com/api/token'
const basicAuthorization = `Basic ${Buffer.from(`${clientId}:${clientSecret}`).toString('base64')}`

// Router
const router = express.Router()
router.use(errors())
router.use(cookieParser(cookieSecret))

/* ==================== */
/* ====== HELPERS ===== */
/* ==================== */

// Generate secure state token
const generateState = () => randomBytes(6).toString('hex')

// Request authorization_code and refresh_token
const requestTokens = (code: string, type: string = 'authorization_code') => {
  const data = {
    grant_type: type,
    redirect_uri: redirectUri,
    code: undefined,
    refresh_token: undefined
  }

  switch (type) {
    case 'refresh_token':
      data.refresh_token = code
      break
    default:
      data.code = code
  }

  const urlEncodedData = stringify(data)

  return axios.post(authorizationUrl, urlEncodedData, {
    headers: {
      Authorization: basicAuthorization
    }
  }).then(((response) => response.data))
}

const constructResponseUrl = (req: any, origin?: string) => {
  // Redirect to origin or app url
  const responseURL = new URL(`${req.protocol}://${req.get('host')}/find`)
  if (origin) {
    responseURL.pathname = decodeURIComponent(origin)
  }

  return responseURL
}

// Handle initial login request (when there is no request token stored)
const handleLogin = (req, res) => {
  // Set origin cookie to redirect back to once the authorization flow is complete
  const { origin } = req.query
  if (origin) {
    res.cookie('origin', origin, { httpOnly: true })
  }

  // Generate state and set in cookie
  const state = generateState()
  res.cookie('spotify_state', state, {
    expires: new Date(Date.now() + 6e4), // a minute
    signed: true,
    httpOnly: true
  })

  // Send Spotify authorization request
  res.redirect('https://accounts.spotify.com/authorize' +
    '?response_type=code' +
    `&client_id=${clientId}` +
    `&scope=${encodeURIComponent(scopes)}` +
    `&state=${state}` +
    `&redirect_uri=${encodeURIComponent(redirectUri)}`
  )
}

// Handle a refresh of the access token using a stored refresh token
const refreshToken = async (req, res, token: string) => {
  // Define origin to redirect back to once the token is refreshed
  const { origin } = req.query

  let accessToken
  try {
    const data = await requestTokens(token, 'refresh_token')
    accessToken = data.access_token
  } catch (err) {
    return res.sendStatus(400)
  }

  // Set access token to expire in an hour
  res.cookie('access_token', accessToken, {
    expires: new Date(Date.now() + 36e5),
    signed: true,
    httpOnly: true
  })

  // Redirect to origin or app url
  const responseURL = constructResponseUrl(req, origin)
  res.redirect(responseURL)
}

/* ==================== */
/* ====== ROUTES ====== */
/* ==================== */

router.get('/login', celebrate({
  signedCookies: Joi.object({
    access_token: Joi.string(),
    refresh_token: Joi.string()
  }).unknown()
}), (req, res) => {
  const { signedCookies: { access_token, refresh_token } } = req

  // Go to the response url if both access_token and refresh_token exist
  if (access_token && refresh_token) {
    const { origin } = req.query
    const responseURL = constructResponseUrl(req, origin)
    res.redirect(responseURL)
  // If there is a refresh_token, but not an access_token, refresh the token
  } else if (refresh_token) {
    return refreshToken(req, res, refresh_token)
  } else {
  // Otherwise handle an initial login
    return handleLogin(req, res)
  }
})

router.get('/redirect', celebrate({
  query: {
    code: Joi.string().required(),
    state: Joi.string().required()
  },
  cookies: Joi.object({
    origin: Joi.string(),
  }).unknown(),
  signedCookies: Joi.object({
    spotify_state: Joi.string().required()
  }).unknown()
}), async (req, res) => {
  // Verify that the states match
  if (req.query.state !== req.signedCookies.spotify_state) {
    return res.status(400).send('states do not match')
  }
  res.clearCookie('spotify_state')

  // Request authorization code and refresh token
  let tokens

  try {
    tokens = await requestTokens(req.query.code)
  } catch (err) {
    return res.status(400)
  }

  // Store tokens in cookies
  const { access_token, refresh_token } = tokens

  // set access token to expire in an hour
  res.cookie('access_token', access_token, {
    expires: new Date(Date.now() + 36e5),
    signed: true,
    httpOnly: true
  })

  // set refresh token to expire in 2 weeks
  res.cookie('refresh_token', refresh_token, {
    expires: new Date(Date.now() + 12096e5),
    signed: true,
    httpOnly: true
  })

  // Redirect to origin or app url
  const origin = req.cookies && req.cookies.origin
  const responseURL = constructResponseUrl(req, origin)

  // Clear the origin cookie
  if (origin) res.clearCookie('origin')

  // Redirect to the response URL
  res.redirect(responseURL)
})

router.get('/logout', (_, res) => {
  // clear user data
  res.clearCookie('access_token')
  res.clearCookie('refresh_token')

  res.redirect('/')
})

export default router
