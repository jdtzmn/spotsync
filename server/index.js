const path = require('path')
const express = require('express')
const app = express()
const server = require('http').Server(app)
const cookieParser = require('cookie-parser')

// security middleware
const RateLimit = require('express-rate-limit')
const helmet = require('helmet')
const csrf = require('csurf')

// ================================
// ------------ TOKENS ------------
// ================================
let tokens = {}
try {
  tokens = require('./secrets.js')
} catch (e) {
  const {
    SPOTIFY_CLIENT_ID,
    SPOTIFY_CLIENT_SECRET,
    COOKIES_SECRET
  } = process.env

  if (!SPOTIFY_CLIENT_ID || !SPOTIFY_CLIENT_SECRET || !COOKIES_SECRET) {
    throw new Error('client_id, client_secret, and cookies_secret are required')
  }
}

const clientId = process.env.SPOTIFY_CLIENT_ID || tokens.client_id
const clientSecret = process.env.SPOTIFY_CLIENT_SECRET || tokens.client_secret
const cookieSecret = process.env.COOKIES_SECRET || tokens.cookies

// ================================
// ---------- MIDDLEWARE ----------
// ================================
app.use(helmet({
  frameguard: { action: 'deny' }
}))
app.use(cookieParser(cookieSecret))
app.use(express.static(path.join(__dirname, '../dist')))
// enable csrf protection if in production
if (process.env.NODE_ENV === 'production') {
  app.use(csrf({ cookie: true }))
}

// --------------------------------
// -------- RATE LIMITING  --------
// --------------------------------

// enable reverse proxy if in production
if (process.env.NODE_ENV === 'production') {
  app.enable('trust proxy')
}

const limiter = new RateLimit({
  windowMs: 10 * 60 * 1000, // 10 minutes
  max: 100,
  delayMs: 0
})

app.use(limiter)

// ================================
// ------------ ROUTES ------------
// ================================

  // --------------------------------
  // -------- AUTHENTICATION  -------
  // --------------------------------
const authRoutes = require('./auth.js')(clientId, clientSecret)
app.use('/auth', authRoutes)

  // --------------------------------
  // ------- SERVE BUILT FILES ------
  // --------------------------------

app.get('*', (req, res) => {
  if (req.csrfToken) {
    res.cookie('XSRF-TOKEN', req.csrfToken())
  }
  res.sendFile(path.join(__dirname, '../dist/index.html'))
})

// ================================
// ---------- WEB SOCKETS ---------
// ================================

require('./sockets.js')(server, cookieSecret)

// --------------------------------
// ------------ LISTEN  -----------
// --------------------------------
server.listen(process.env.PORT || 3000, () => {
  console.log('Listening on port:', server.address().port)
})
