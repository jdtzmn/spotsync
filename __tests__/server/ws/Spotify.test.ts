/**
 * @jest-environment node
 */

import Spotify from 'server/ws/Spotify'
import { clientId, clientSecret } from 'server/config'

describe('Spotify', () => {
  let spotify: Spotify
  beforeEach(() => {
    spotify = new Spotify(clientId, clientSecret)
  })

  it('should initialize', () => {
    expect(spotify).toBeDefined()
  })

  it('should setup and set isSetup property', () => {
    expect(spotify.isSetup).toBeFalsy()
    return spotify.setup()
      .then(() => {
        expect(spotify.isSetup).toBeTruthy()
      })
  })

  it('should return 400 on invalid track', () => {
    return spotify.getTrack('invalid_uri')
      .catch((err) => {
        expect(err.response.status).toBe(400)
      })
  })

  it('should return correct spotify track data on request', async () => {
    // Hey Brother by Avicii
    return spotify.getTrack('7tFiyTwD0nx5a1eklYtX2J')
      .then((trackData) => expect(trackData).toMatchSnapshot())
  })
})
