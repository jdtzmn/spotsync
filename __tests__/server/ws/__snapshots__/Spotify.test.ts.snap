// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spotify should return correct spotify track data on request 1`] = `
Object {
  "album": Object {
    "album_type": "album",
    "artists": Array [
      Object {
        "external_urls": Object {
          "spotify": "https://open.spotify.com/artist/1dfeR4HaWDbWqFHLkxsg1d",
        },
        "href": "https://api.spotify.com/v1/artists/1dfeR4HaWDbWqFHLkxsg1d",
        "id": "1dfeR4HaWDbWqFHLkxsg1d",
        "name": "Queen",
        "type": "artist",
        "uri": "spotify:artist:1dfeR4HaWDbWqFHLkxsg1d",
      },
    ],
    "available_markets": Array [
      "CA",
      "US",
    ],
    "external_urls": Object {
      "spotify": "https://open.spotify.com/album/6X9k3hSsvQck2OfKYdBbXr",
    },
    "href": "https://api.spotify.com/v1/albums/6X9k3hSsvQck2OfKYdBbXr",
    "id": "6X9k3hSsvQck2OfKYdBbXr",
    "images": Array [
      Object {
        "height": 640,
        "url": "https://i.scdn.co/image/f6e02ce8bc9e91afa0e220703072c20bf72c72c5",
        "width": 640,
      },
      Object {
        "height": 300,
        "url": "https://i.scdn.co/image/f4111b2cd4d2459cfdf9aad9010d74c97b7e3a68",
        "width": 300,
      },
      Object {
        "height": 64,
        "url": "https://i.scdn.co/image/8ea6edc8b88d8777c931bff14f48d70503de674c",
        "width": 64,
      },
    ],
    "name": "A Night At The Opera (Deluxe Remastered Version)",
    "release_date": "1975-11-21",
    "release_date_precision": "day",
    "total_tracks": 18,
    "type": "album",
    "uri": "spotify:album:6X9k3hSsvQck2OfKYdBbXr",
  },
  "artists": Array [
    Object {
      "external_urls": Object {
        "spotify": "https://open.spotify.com/artist/1dfeR4HaWDbWqFHLkxsg1d",
      },
      "href": "https://api.spotify.com/v1/artists/1dfeR4HaWDbWqFHLkxsg1d",
      "id": "1dfeR4HaWDbWqFHLkxsg1d",
      "name": "Queen",
      "type": "artist",
      "uri": "spotify:artist:1dfeR4HaWDbWqFHLkxsg1d",
    },
  ],
  "available_markets": Array [
    "CA",
    "US",
  ],
  "disc_number": 1,
  "duration_ms": 354320,
  "explicit": false,
  "external_ids": Object {
    "isrc": "GBUM71029604",
  },
  "external_urls": Object {
    "spotify": "https://open.spotify.com/track/7tFiyTwD0nx5a1eklYtX2J",
  },
  "href": "https://api.spotify.com/v1/tracks/7tFiyTwD0nx5a1eklYtX2J",
  "id": "7tFiyTwD0nx5a1eklYtX2J",
  "is_local": false,
  "name": "Bohemian Rhapsody - Remastered 2011",
  "popularity": 79,
  "preview_url": null,
  "track_number": 11,
  "type": "track",
  "uri": "spotify:track:7tFiyTwD0nx5a1eklYtX2J",
}
`;
